### PR TITLE
Cpp dbg

### PR DESCRIPTION
**Pull request by jleidel**
Merging the following functionality: 
- better paging support
- fixes for nearly all compressed instructions
- fixes for decoding compressed instructions
- fixes for large .BSS sections
- misc fixes for jump instructions
- CMake fixes
- better debugging support

### CODE REVIEW

File: include/insns/RV32I.h [hunk 1]
```diff
@@ -79,7 +81,9 @@ namespace SST{
                      RevMem *M, RevInst Inst) {
         // c.j $imm = jal x0, $imm
         Inst.rd = 0; // x0
-        Inst.imm = Inst.jumpTarget;
+        //Inst.imm = Inst.jumpTarget;
+        SEXT(Inst.imm, Inst.jumpTarget, 12);
+        std::cout << "executing C.J : Imm=0x" << std::hex << Inst.imm << std::dec << std::endl;
```
Comment (donofrio) [inline include/insns/RV32I.h:93]: Remove spurious cout or wrap in #ifdef

File: include/insns/RV64I.h [hunk 2]
```diff
@@ -104,6 +106,10 @@ namespace SST{
                     REVMEM_FLAGS(0x00));
         R->cost += M->RandCost(F->GetMinCost(),F->GetMaxCost());
         R->RV64_PC += Inst.instSize;
+        if( R->RV64[Inst.rs1]+(int32_t)(td_u32(Inst.imm,12)) == 0x88d08 ){
+          std::cout << "Loading from 0x88d08" << std::endl;
+          std::cout << "x[" << Inst.rd << "] = 0x" << std::hex << R->RV64[Inst.rd] << std::dec << std::endl;
+        }
```
Comment (donofrio) [inline include/insns/RV64I.h:115]: Remove address check and print

File: src/RevPrefetcher.cc [hunk 3]
```diff
@@ -12,9 +12,10 @@
 
 bool RevPrefetcher::IsAvail(uint64_t Addr){
 
-  // note: this logic does not consider compressed instructions
-  //       however, it should work for what we need.  we only need
-  //       to know whether an instruction as been filled
+  // note: this logic now considers compressed instructions
+  if( Addr == 0x4bd38ull ){
```
Comment (donofrio) [inline src/RevPrefetcher.cc:16]: remove address check

File: src/RevPrefetcher.cc [hunk 4]
```diff
@@ -38,6 +39,10 @@ bool RevPrefetcher::IsAvail(uint64_t Addr){
         return false;
       }
 
+      if( Addr == 0x4bd38ull ){
+        std::cout << "not stalled" << std::endl;
+      }
+
```
Comment (donofrio) [inline src/RevPrefetcher.cc:45]: remove address check

File: src/RevPrefetcher.cc [hunk 5]
```diff
@@ -58,6 +63,10 @@ bool RevPrefetcher::IsAvail(uint64_t Addr){
     }
   }
 
+  if( Addr == 0x4bd38ull ){
+    std::cout << "Filling 0x4bd38" << std::endl;
+  }
+
   // if we reach this point, then the instruction hasn't even triggered
```
Comment (donofrio) [inline src/RevPrefetcher.cc:70]: remove address check

File: src/RevPrefetcher.cc [hunk 6]
```diff
@@ -165,8 +174,16 @@ bool RevPrefetcher::InstFetch(uint64_t Addr, bool &Fetched, uint32_t &Inst){
 
 void RevPrefetcher::Fill(uint64_t Addr){
 
+  if( Addr == 0x4bd38ull ){
+    std::cout << "inside Fill" << std::endl;
+  }
+
   // determine if the address is 32bit aligned
```
Comment (donofrio) [inline src/RevPrefetcher.cc:181]: remove address check or wrap in #ifdef

File: src/RevPrefetcher.cc [hunk 7]
```diff
@@ -165,8 +174,16 @@ bool RevPrefetcher::InstFetch(uint64_t Addr, bool &Fetched, uint32_t &Inst){
...
   if((Addr%4)!=0){
+    if( Addr == 0x4bd38ull ){
+      std::cout << "address not aligned" << std::endl;
+      std::cout << "filling at 0x" << std::hex << (Addr&0xFFFFFFFFFFFFFFFC) << std::dec << std::endl;
+    }
```
Comment (donofrio) [inline src/RevPrefetcher.cc:186]: remove address check or wrap in #ifdef

File: src/RevPrefetcher.cc (hunk 6 above)
Comment (jleidel) [reply src/RevPrefetcher.cc]: resolved in https://github.com/tactcomplabs/rev/commit/e9a5df3d3de81136677d389638c82e5ced6b65c9

File: src/RevPrefetcher.cc (hunk 5 above)
Comment (jleidel) [reply src/RevPrefetcher.cc]: resolved in https://github.com/tactcomplabs/rev/commit/e9a5df3d3de81136677d389638c82e5ced6b65c9

File: src/RevPrefetcher.cc (hunk 4 above)
Comment (jleidel) [reply src/RevPrefetcher.cc]: resolved in https://github.com/tactcomplabs/rev/commit/e9a5df3d3de81136677d389638c82e5ced6b65c9

File: src/RevPrefetcher.cc (hunk 3 above)
Comment (jleidel) [reply src/RevPrefetcher.cc]: resolved in https://github.com/tactcomplabs/rev/commit/e9a5df3d3de81136677d389638c82e5ced6b65c9

File: include/insns/RV64I.h (hunk 2 above)
Comment (jleidel) [reply include/insns/RV64I.h]: resolved in https://github.com/tactcomplabs/rev/commit/e9a5df3d3de81136677d389638c82e5ced6b65c9

**Comment by rkabrick** [comment]
I added a commit that prevents register x0 from ever getting set as a dependency. This was a bug I encountered during the system call work as it affects `ecall` instructions but until that stuff gets merged I figured it'd be worth fixing now.

**Pull request merged**